LightAPRS-hab/LightAPRS-hab.ino

```diff
@@ -106,90 +171,201 @@ void setup() {
-
-    if ((gps.location.age() < 1000 || gps.location.isUpdated()) && gps.location.isValid()) {
-      if (gps.satellites.isValid() && (gps.satellites.value() > 3)) {
-      updatePosition();
-      updateTelemetry();
-      
-      //GpsOFF;
-      setGPS_PowerSaveMode();
-      //GpsFirstFix=true;
+    }
 
-      if(autoPathSizeHighAlt && gps.altitude.feet()>3000){
+    if(secsTillTx <= 0) {
+      last_tx_millis = millis();
+      secsTillTx = GPSWait;
+
+      updateGpsData(1000);
+      gpsDebug();
+
+      if ((gps.location.age() < 1000 || gps.location.isUpdated()) && 
+          gps.location.isValid()) {
+        
+        if (gps.satellites.isValid() && 
+           gps.satellites.value() > 3) {
+
+          doAltCheck();
+          updateZone(); //updates BeaconWait
+          updateComment();
+          updatePosition();
+          updateTelemetry();
+          secsTillTx = BeaconWait;
+
+          //GpsOFF;
+          setGPS_PowerSaveMode();
+          //GpsFirstFix=true;
+          current_altitude = gps.altitude.feet();
+          if(current_altitude > max_altitude) {
+            max_altitude = current_altitude;
+          }
+          if(autoPathSizeHighAlt && gps.altitude.feet() > 3000){
             //force to use high altitude settings (WIDE2-n)
             APRS_setPathSize(1);
-        } else {
+          } else {
             //use defualt settings  
             APRS_setPathSize(pathSize);
-        }
-      
-      //send status message every 60 minutes
-      if(gps.time.minute() == 30){               
-        sendStatus();       
+          }
+          
+          //send status message every 15 minutes
+          if((gps.time.minute() % 15) == 0) {               
+            sendStatus();       
+          } else {
+            sendLocation();
+          }
+
+          freeMem();
+          Serial.flush();
+        } // if time to tx
+        // sleepSeconds(BeaconWait-((millis-loop_start)/1000));
       } else {
+#if defined(DEVMODE)
+      Serial.println(F("Not enough satelites"));
+#endif
+      }
+    }
 
-         sendLocation();
+    secsTillTx -= round((millis()-loop_start)/1000);
+  } else {
+    secsToCheckBatt--;
 
-      }
+    secsToCheckBatt -= (millis()-loop_start)/1000;
+    // sleepSeconds(BattWait-((millis-loop_start)/1000));
+  }
+  // Serial.println("Loop time in milliseconds->");
+  // Serial.println(round((millis()-loop_start)/1000));
+  sleepSeconds(secsTillTx);
+  secsTillTx = 0;
+}
 
-      freeMem();
-      Serial.flush();
-      sleepSeconds(BeaconWait);
 
-      } else {
-#if defined(DEVMODE)
-      Serial.println(F("Not enough sattelites"));
-#endif
-      }
-    } 
+
+// Begin smartPacket Functions
+void doAltCheck() {
+  if (gps.altitude.feet() <= lastalt) {
+    balloonDescendRepeat ++;
   } else {
+    balloonDescendRepeat = 0;
+  }
+  if (balloonDescendRepeat >= numDescendChecks) {
+    balloonPopped = true;
+  }
+}
+void updateZone() {
+  
+  long a = gps.altitude.feet();
```

Comment:
is there a performance cost to calling this function or is it just reading from memory?  if it costs in performance, we need to read it once into a variable, then use that variable.

Reply:
There is a veeeery small performance cost.